File: java/console/src/main/java/com/lacunasoftware/signer/sample/scenarios/CreateDocumentWithAttachmentScenario.java
package com.lacunasoftware.signer.sample.scenarios;

import com.lacunasoftware.signer.FileUploadModel;
import com.lacunasoftware.signer.FlowActionType;
import com.lacunasoftware.signer.attachments.AttachmentUploadModel;
import com.lacunasoftware.signer.documents.CreateDocumentRequest;
import com.lacunasoftware.signer.documents.CreateDocumentResult;
import com.lacunasoftware.signer.flowactions.FlowActionCreateModel;
import com.lacunasoftware.signer.javaclient.builders.AttachmentUploadBuilder;
import com.lacunasoftware.signer.javaclient.builders.FileUploadModelBuilder;
import com.lacunasoftware.signer.javaclient.exceptions.RestException;
import com.lacunasoftware.signer.javaclient.models.UploadModel;
import com.lacunasoftware.signer.sample.Util;
import com.lacunasoftware.signer.users.ParticipantUserModel;

import java.io.IOException;
import java.util.ArrayList;

public class CreateDocumentWithAttachmentScenario extends Scenario {
    /**
     * This scenario demonstrates the creation of a document with Attachment.
     */
    @Override
    public void Run() throws IOException, RestException {
        // 1. The file's bytes must be read by the application and uploaded
        byte[] content = Util.getInstance().getResourceFile("sample.pdf");
        UploadModel uploadModel = signerClient.uploadFile("sample.pdf", content, "application/pdf");

        // 2. Define the name of the document which will be visible in the application
        FileUploadModelBuilder fileUploadModelBuilder = new FileUploadModelBuilder(uploadModel);
        fileUploadModelBuilder.setDisplayName("One Signer Sample");


        // 3. Repeat the same steps above but now for the attachment and using AttachmentUploadBuilder
        byte[] attachmentContent = Util.getInstance().getResourceFile("sample.pdf");
        UploadModel attachmentModel = signerClient.uploadFile("sample.pdf", attachmentContent, "application/pdf");
        AttachmentUploadBuilder attachmentFileUploadModelBuilder = new AttachmentUploadBuilder(attachmentModel);
        attachmentFileUploadModelBuilder.setDisplayName("One Attachment Sample");
        //here you can define if the attachment will be restrict to the organization participants (true) or to everyone in the document flow (false)
        attachmentFileUploadModelBuilder.setIsPrivate(false);


        // 4. For each participant on the flow, create one instance of ParticipantUserModel
        ParticipantUserModel user = new ParticipantUserModel();
        user.setName("Jack Bauer");
        user.setEmail("jack.bauer@mailinator.com");
        user.setIdentifier("75502846369");

        // 5. Create a FlowActionCreateModel instance for each action
        FlowActionCreateModel flowActionCreateModel = new FlowActionCreateModel();
        flowActionCreateModel.setType(FlowActionType.SIGNER);
        flowActionCreateModel.setUser(user);

        // 6. Send the document create request with attachment's attribute
        CreateDocumentRequest documentRequest = new CreateDocumentRequest();
        documentRequest.setFiles(new ArrayList<FileUploadModel>() {
            private static final long serialVersionUID = 1L;
            {
                add(fileUploadModelBuilder.toModel());
            }
        });
        documentRequest.setAttachments(new ArrayList<AttachmentUploadModel>(){
            private static final long serialVersionUID = 1L;
            {
                add(attachmentFileUploadModelBuilder.toModel());
            }
        });
        documentRequest.setFlowActions(new ArrayList<FlowActionCreateModel>() {
            private static final long serialVersionUID = 1L;
            {
                add(flowActionCreateModel);
            }
        });
        CreateDocumentResult result = signerClient.createDocument(documentRequest).get(0);

        System.out.println(String.format("Document %s created", result.getDocumentId().toString()));
    }

}

File: nodejs/scenarios/createDocumentWithAttachmentScenario.ts
import {
  UploadApi,
  DocumentsApi,
  FileUploadModel,
  UsersParticipantUserModel,
  FlowActionsFlowActionCreateModel,
  FlowActionType,
  getBase64,
  AttachmentsAttachmentUploadModel,
} from "signer-node-client";
import { config } from "./scenario";

/**
 * This scenario demonstrates the creation of a document with Attachment.
 */

const filePath = "../../samples/sample.pdf";
const fileName = "sample.pdf";

const uploadApi = new UploadApi(config);
const documentsApi = new DocumentsApi(config);

// 1. The file's bytes must be read by the application and uploaded
uploadApi
  .apiUploadsBytesPost({
    bytes: getBase64(filePath),
  })
  .then((res) => {
    // 2. Define the name of the document which will be visible in the application
    const fileUploadModel: FileUploadModel = {
      displayName: "Document sample",
      id: res.data.id,
      contentType: "application/pdf",
      name: fileName,
    };

    // 3. Repeat the same steps above but now for the attachment and using AttachmentUploadBuilder
    uploadApi
      .apiUploadsBytesPost({
        bytes: getBase64(filePath),
      })
      .then((res) => {
        const attachmentFileUploadModelBuilder: AttachmentsAttachmentUploadModel =
          {
            id: res.data.id,
            name: fileName,
            contentType: "application/pdf",
            displayName: "One Attachment Sample",
            //here you can define if the attachment will be restrict to the organization participants (true) or to everyone in the document flow (false)
            isPrivate: false,
          };

        // 4. For each participant on the flow, create one instance of ParticipantUserModel
        const participant: UsersParticipantUserModel = {
          name: "Jack Bauer",
          email: "jack.bauer@mailnator.com",
          identifier: "75502846369",
        };

        // 5. Create a FlowActionCreateModel instance for each action
        const flowAction: FlowActionsFlowActionCreateModel = {
          type: FlowActionType.Signer,
          user: participant,
        };

        // 6. Send the document create request with attachment's attribute
        documentsApi
          .apiDocumentsPost({
            files: [fileUploadModel],
            // set the attachment
            attachments: [attachmentFileUploadModelBuilder],
            flowActions: [flowAction],
          })
          .then((res) => {
            // 7. Print result
            console.log("Document ", res.data[0].documentId, "created");
          });
      });
  });
